cloudformation/lambda/AutoOffEC2.yaml

```diff
@@ -0,0 +1,104 @@
+                      filters = [
+                          {
+                              'Name': 'instance-state-name',
+                              'Values': ['running']
+                          }
+                      ]
+
+                      # filter the instances
+                      instances = ec2.instances.filter(Filters=filters)
+                      stop_instances = set()
+                      prod_instances = {instance.id if prod_instance(instance) else stop_instances.add(instance.id) for instance in
+                                        instances}
+
+                      logger.info("All Instances")
+                      all_instances = prod_instances.union(stop_instances)
+                      logger.info(len(all_instances))
+                      logger.info(all_instances)
+
+                      logger.info("Production Instances")
+                      logger.info(len(prod_instances))
+                      logger.info(prod_instances)
+
+                      logger.info("Stopping Instances")
+                      logger.info(len(stop_instances))
+                      logger.info(stop_instances)
+
+                      if len(stop_instances) > 0:
+                          # perform the shutdown
+                          logger.info("Shutting down instances")
+                          logger.info(stop_instances)
+                          shutting_down = ec2.instances.filter(InstanceIds=stop_instances).stop()
+                          logger.info(shutting_down)
+
+
+                  def prod_instance(instance):
+                      if instance.tags is not None:
+                          for tag in instance.tags:
+                              logger.info(tag)
+                              if "Name" in tag['Key'] and "-aws-syd" in tag['Value']:
+                                  return True
+                              elif "elasticbeanstalk" in tag['Key']:
+                                  return True
+                              elif "AutoOff" in tag['Key'] and "False" in tag['Value']:
+                                  return True
+                          return False
+
+  LambdaExecutionRole:
+      Type: AWS::IAM::Role
+      Properties:
+          AssumeRolePolicyDocument:
+              Version: '2012-10-17'
+              Statement:
+              - Effect: Allow
+                Principal: {Service: [lambda.amazonaws.com]}
+                Action: ['sts:AssumeRole']
+          Path: /
+          ManagedPolicyArns:
+          - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
+          - arn:aws:iam::aws:policy/CloudWatchLogsFullAccess
+          Policies:
+                - PolicyName: PublishAndExecute
+                  PolicyDocument:
+                    Version: 2012-10-17
+                    Statement:
+                    - Effect: Allow
+                      Action: ['ec2:StopInstances', 'lambda:PublishVersion']
+                      Resource: '*'
```

Comment:
Would be great if we could limit the permissions to only the instances actually tagged as AutoOff, rather than giving the code full access and relying on that to not do the wrong thing.

See the "conditions" stuff in here, and just replace with the AutoOff tag for matching:

https://aws.amazon.com/premiumsupport/knowledge-center/iam-ec2-resource-tags/